Relative color syntax doesn't work with currentColor or system colors

weinig opened the pull request:
#### c05a19a495bdffa90ca2c5707c651dee249b6870
<pre>
Relative color syntax doesn&apos;t work with currentColor or system colors
<a href="https://bugs.webkit.org/show_bug.cgi?id=245970">https://bugs.webkit.org/show_bug.cgi?id=245970</a>
<a href="https://rdar.apple.com/problem/100981965">rdar://problem/100981965</a>

Reviewed by Darin Adler.

Adds support for late resolution of relative colors, matching the
model set by color-mix.

Re-works color parsing to completely split resolution from consuming,
allowing for a single implementation of the parsing code, all returning
CSSUnresolvedColor values. This allows us to delay resolving relative
colors until needed, storing the components first as a CSSUnresolvedRelativeColor
and then a StyleRelativeColor (mirroring CSSUnresolvedColorMix and
StyleColorMix).

* LayoutTests/TestExpectations:
    - Removes skips for now passing currentColor tests.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color.html:
    - Added tests for currentColor at computed value time.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color.html:
    - Updated tests to match spec (and new implementation) which preserves
      declared values for declared value serialization of relative colors.

* LayoutTests/imported/w3c/web-platform-tests/css/support/color-testcommon.js:
    - Add helper which allows fuzzy testing using a specific property, rather
      than always hardcoding it to use &quot;color&quot;.

* Source/WTF/wtf/VectorTraits.h:
    - Add UniqueRef to the set of smart pointers which get the simple class
      traits, needed to use UniqueRef with FixedVector.

* Source/WTF/wtf/text/TextStream.h:
(WTF::streamSizedContainer):
(WTF::operator&lt;&lt;):
    - Extract sized container serialization into a helper and use to replace
      duplicated implementations. Also add overload for FixedVector.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
    - Expose number serialization helper to allow identical serialization when
      not using CSSPrimitiveValue for numeric values.

* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::buildCSSTextRecursive):
    - Ensure that symbols get serialized when they are part of a calc().

* Source/WebCore/css/calc/CSSCalcSymbolNode.cpp:
* Source/WebCore/css/calc/CSSCalcSymbolNode.h:
(WebCore::CSSCalcSymbolNode::customCSSText const):
    - Expose serialization of a symbol node utilizing `nameLiteralForSerialization`.

* Source/WebCore/css/color/CSSAbsoluteColorResolver.h: Added.
    - Factor out code to resolve a parsed absolute color into a concrete Color.

* Source/WebCore/css/color/CSSColorConversion+ToColor.h: Added.
    - Factor out code to convert a &quot;typed color&quot; into a concrete Color, applying
      the rules specified by the descriptor provided.

* Source/WebCore/css/color/CSSColorConversion+ToTypedColor.h: Added.
    - Factor out code to convert a &quot;parsed color&quot; into a &quot;typed color&quot;, normalizing
      values according to the color descriptor provided.

* Source/WebCore/css/color/CSSColorDescriptors.h:
    - Moves descriptors out of CSSPropertyParserHelpers, they are now used outside
      of that area.
    - Add additional helpers for working with the descriptors that compute the
      component types at various stages, such as when calc and raw symbols have
      not yet been evaluated.
    - Add additional per-descriptor data to support serialization of relative
      color functions that have not been fully resolved.

* Source/WebCore/css/color/CSSColorMixResolver.cpp: Renamed from Source/WebCore/css/color/CSSResolvedColorMix.cpp.
* Source/WebCore/css/color/CSSColorMixResolver.h: Renamed from Source/WebCore/css/color/CSSResolvedColorMix.h.
    - Renamed to follow the same pattern we are using with relative color resolution.

* Source/WebCore/css/color/CSSColorMixSerialization.cpp: Copied from Source/WebCore/css/color/CSSColorMixSerialization.h.
* Source/WebCore/css/color/CSSColorMixSerialization.h:
    - Moved a few non-template functions into an implementation file to avoid
      unnecessary inlining for non-perf critical code.

* Source/WebCore/css/color/CSSRelativeColorResolver.h: Added.
    - Factor out code to resolve a parsed relative color with its origin already
      resolved into a concrete Color. Resolution is performed in stages:

        1. If any components are raw symbols, replace them with a their value
           in the symbol table.
        2. If any components are `calc()` values, evaluate the `calc()` passing
           the symbol table for any symbolic values they retain.
        3. At this point, all the components are raw values of some type (number,
           percent, or angle), normalize those raw components into a typed color
           (e.g. SRGBA&lt;float&gt;, OKLab&lt;float&gt;, etc.).
        4. Wrap the typed color in the type-erased wrapper, Color.

* Source/WebCore/css/color/CSSRelativeColorSerialization.cpp: Added.
* Source/WebCore/css/color/CSSRelativeColorSerialization.h: Added.
    - Factor out code to serialize relative colors.

* Source/WebCore/css/color/CSSUnresolvedAbsoluteColor.cpp: Added.
* Source/WebCore/css/color/CSSUnresolvedAbsoluteColor.h: Added.
    - Type to represent fully resolved colors that are part of CSSUnresolvedColor tree.
      Currently used for absolute color values and the soon to be removed color-contrast().

* Source/WebCore/css/color/CSSUnresolvedColor.cpp:
* Source/WebCore/css/color/CSSUnresolvedColor.h:
    - Add all remaining color types to the value variant to represent declared color
      values. To retain the enough information for correct resolution of relative
      colors, the variant holds a fully specified CSSUnresolvedRelativeColor for each
      relative color supported. Hex and Keyword types were also added to allow removing
      use of CSSPrimitiveValue in the CSSUnresolvedColor tree. Now, child colors are
      represented as UniqueRef&lt;CSSUnresolvedColor&gt;.
    - The new createColor() function allows eager resolution of unresolved colors as
      needed by the `parseColorRaw()` functions that Canvas uses.

* Source/WebCore/css/color/CSSUnresolvedColorHex.cpp: Added.
* Source/WebCore/css/color/CSSUnresolvedColorHex.h: Added
    - Type to represent hex colors in the CSSUnresolvedColor tree.

* Source/WebCore/css/color/CSSUnresolvedColorKeyword.cpp: Added.
* Source/WebCore/css/color/CSSUnresolvedColorKeyword.h: Added.
    - Type to represent keyword colors in the CSSUnresolvedColor tree.

* Source/WebCore/css/color/CSSUnresolvedColorMix.cpp:
* Source/WebCore/css/color/CSSUnresolvedColorMix.h:
    - Re-worked to remove use of CSSPrimitiveValue representations of mix components,
      instead using `UniqueRef&lt;CSSUnresolvedColor&gt;` for the origin colors and
      `std::variant&lt;PercentRaw, UnevaluatedCalc&lt;PercentRaw&gt;&gt;` for the percentages.
      This preserves the type information and makes these usable in workers without
      extensive CSSValuePool piping.

* Source/WebCore/css/color/CSSUnresolvedColorResolutionContext.cpp: Added.
* Source/WebCore/css/color/CSSUnresolvedColorResolutionContext.h: Added.
    - Context used by the `parseColorRaw()` functions to resolve colors without
      going through the style system. Currently the context contains explicit
      values, but if pre-computing these turns out of be expensive, this can
      be converted into a delegate to allow lazy evaluation.

* Source/WebCore/css/color/CSSUnresolvedLightDark.cpp:
* Source/WebCore/css/color/CSSUnresolvedLightDark.h:
    - Like CSSUnresolvedColorMix, re-worked to remove use of CSSPrimitiveValue.

* Source/WebCore/css/color/CSSUnresolvedRelativeColor.cpp: Added.
* Source/WebCore/css/color/CSSUnresolvedRelativeColor.h: Added.
    - Type to represent relative color in the CSSUnresolvedColor tree.
    - Uses a color descriptor to define component types and stores the
      origin as a UniqueRef&lt;CSSUnresolvedColor&gt;.
    - Converts to the new StyleRelativeColor at style resolution time
      allowing a currentColor origin value to be preserved as late as
      possible.

* Source/WebCore/css/color/StyleColor.cpp:
* Source/WebCore/css/color/StyleColor.h:
    - Like CSSUnresolvedColor, the relative color types are added to the
      the kind variant, fully specified by color descriptor. One slightly
      unfortunate duplication is that all the relative colors require their
      own constructor functions in order to keep instantiation of the variant
      to the implementation file.
    - Like StyleColorMix, the StyleRelativeColor types are stored in UniqueRefs
      to avoid bloating the size for the relatively uncommon case of currentColor
      as part of the origin.
    - Also like StyleColorMix, the StyleRelativeColor types are first checked
      to see if they can be resolved, and if so, are stored as absolute colors
      instead.

* Source/WebCore/css/color/StyleColorMix.cpp:
* Source/WebCore/css/color/StyleColorMix.h:
    - Use PercentRaw for the Percentage to allow for clearer serialization.
    - Factor out check if the origin is absolute into a function to match
      relative color.
    - Sort functions to match other style color types.

* Source/WebCore/css/color/StyleCurrentColor.cpp:
* Source/WebCore/css/color/StyleCurrentColor.h:
    - Sort functions to match other style color types.

* Source/WebCore/css/color/StyleRelativeColor.h: Added.
    - Type to represent relative colors in the StyleColor tree.
    - Utilizes code shared with CSSUnresolvedRelativeColor for resolution
      and serialization.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.h:
    - Major rework of the parsing logic. Big takeaways:

        - Only does the consuming now, resolution has all been moved to
          other files.
        - Only one set of consumer functions, all working to return a
          CSSUnresolvedColor value. Now more separate normal vs raw
          distinction outside of exposed functions. The exposed functions
          consume the CSSUnresolvedColor, and resolve it to either
          a CSSPrimitiveValue or a raw Color.
        - ColorInterpolationMethod consuming was moved to its own
          file to continue to split things up.
        - Exposed functions have been merged where possible, with all
          now taking CSSColorParsingOptions, and all the raw functions
          also taking a CSSUnresolvedColorResolutionContext.
        - All the functions are now equally worker safe (or at least
          as worker safe as the one labeled worker safe used to be),
          so the worker safe exposed function has been removed.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorInterpolationMethod.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorInterpolationMethod.h: Added.
    Moved from CSSPropertyParserConsumer+Color.h/cpp.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+RawTypes.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+RawTypes.h:
    - Added serialization of raw types, used generically by the relative color
      and color mix serialization code.
    - Removed a number of FooOrBarRaw type definitions that were only used by
      the old color parsing code.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+UnevaluatedCalc.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UnevaluatedCalc.h:
    - Added serialization of UnevaluatedCalc, used generically by the relative color
      and color mix serialization code.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeBorderColor):
(WebCore::CSSPropertyParserHelpers::consumeFontPaletteValuesOverrideColors):
    - Switched to use the CSSColorParsingOptions when calling consumeColor allowing us
      to remove overload.

* Source/WebCore/css/process-css-properties.py:
    - Switched to use the CSSColorParsingOptions when calling consumeColor allowing us
      to remove overload.

* Source/WebCore/html/canvas/CanvasStyle.cpp:
    - Switched to use the plain parseColorRaw with &quot;System&quot; colors not in the allow set,
      rather than worker safe variant, since they are both the same now.
    - Removed hsl clamp opt setting, as that has been a no-op for a bit now.

* Source/WebCore/platform/graphics/ColorSerialization.cpp:
* Source/WebCore/platform/graphics/ColorSerialization.h:
    - Expose serialization of the color space names for use by relative color serialization.

* Source/WebCore/style/ColorFromPrimitiveValue.cpp:
* Source/WebCore/style/ColorFromPrimitiveValue.h:
    - Factor out conversion from CSS value keyword to StyleColor into its own function
      for use by CSSUnresolvedColorKeyword.

Canonical link: <a href="https://commits.webkit.org/278696@main">https://commits.webkit.org/278696@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b712f8193bf53b72e8d7f5a2631881875230b99c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51328 "24 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30634 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3670 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54591 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2018 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53631 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36947 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1697 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/41820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53427 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28281 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44258 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/22938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/25604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1516 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/44669 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/47574 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1589 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56181 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/50830 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26443 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/1484 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/51347 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27686 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44321 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11228 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28576 "Built successfully") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/63154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27421 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/63154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/VectorTraits.h:
```diff
@@ -20,11 +20,12 @@
 #include <memory>
 #include <type_traits>
+#include <utility>
+#include <wtf/Ref.h>
```
Since `RefPtr.h` includes `Ref.h`, there is no need to include `Ref.h`. (I know you just re-sorted here.)

darinadler commented on Source/WTF/wtf/text/TextStream.h:
```diff
@@ -240,26 +240,50 @@ TextStream& operator<<(TextStream& ts, const Markable<T, Traits>& item)
 
-template<typename ItemType, size_t inlineCapacity>
-TextStream& operator<<(TextStream& ts, const Vector<ItemType, inlineCapacity>& vector)
+template<typename SizedContainer>
```
Another way to do this is to define `SizedContainer` as a concept, and then use `template<SizedContainer SizedContainerType>`.

weinig replied:
I played with this, but couldn't get a concept that worked in all cases (kept getting compile errors about DataSegmentVectorEntry use with TextStream). Will give it another try once this lands.

darinadler commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -1212,24 +1212,43 @@ String CSSPrimitiveValue::stringValue() const
+
+static NEVER_INLINE void formatNonfiniteValue(StringBuilder& builder, double number, ASCIILiteral suffix)
+{
+    return builder.append(formatNonfiniteValuePrefix(number), suffix.isEmpty() ? ""_s : " * 1"_s, suffix);
```
Kinda wish we could share the argument list between the `StringBuilder` and `makeString` versions.

darinadler commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -1212,24 +1212,43 @@ String CSSPrimitiveValue::stringValue() const
-    if (!std::isfinite(m_value.number))
-        return formatNonfiniteValue(m_value.number, suffix);
-    return makeString(FormattedCSSNumber::create(m_value.number), suffix);
+    return CSSPrimitiveValue::formatNumberValue(m_value.number, suffix);
```
No need for the `CSSPrimitiveValue::` prefix here.

darinadler commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -1212,24 +1212,43 @@ String CSSPrimitiveValue::stringValue() const
+    return makeString(formatNonfiniteValuePrefix(number), suffix.isEmpty() ? ""_s : " * 1"_s, suffix);
+}
+
+NEVER_INLINE void CSSPrimitiveValue::formatNumberValue(StringBuilder& builder, double value, ASCIILiteral suffix)
```
Does this need to be a member function?

weinig replied:
Nope. And I am planning on moving all this to a new CSSUnit specific home soon.

darinadler commented on Source/WebCore/css/CSSPrimitiveValue.cpp:
```diff
@@ -1212,24 +1212,43 @@ String CSSPrimitiveValue::stringValue() const
+    return builder.append(FormattedCSSNumber::create(value), suffix);
+}
+
+NEVER_INLINE String CSSPrimitiveValue::formatNumberValue(double value, ASCIILiteral suffix)
```
Does this need to be a member function?

weinig replied:
Nope.

darinadler commented on Source/WebCore/css/calc/CSSCalcSymbolNode.h:
```diff
@@ -29,10 +29,14 @@
 
 namespace WebCore {
 
+enum CSSValueID : uint16_t;
```
Why are we adding this in this patch?

weinig replied:
Compile error. Something new was now including this and there was no definition for CSSValueID.

darinadler commented on Source/WebCore/css/color/CSSColorConversion+ToColor.h:
```diff
@@ -0,0 +1,131 @@
+    return unresolved.whiteness < 0.0 || unresolved.whiteness > 100.0 || unresolved.blackness < 0.0 || unresolved.blackness > 100.0;
+}
+
+inline bool outsideSRGBGamut(SRGBA<float>)
```
Maybe even `constexpr`, not just `inline`?

darinadler commented on Source/WebCore/css/color/CSSUnresolvedAbsoluteColor.cpp:
```diff
@@ -0,0 +1,71 @@
+bool containsCurrentColor(const CSSUnresolvedAbsoluteColor&)
+{
+    return false;
+}
```
Maybe LTO takes care of this, but I think you’d want these to be inlined.

weinig replied:
I can also make them constexpr in the header.

Commit:
Relative color syntax doesn't work with currentColor or system colors
https://bugs.webkit.org/show_bug.cgi?id=245970
rdar://problem/100981965

Reviewed by Darin Adler.

Adds support for late resolution of relative colors, matching the
model set by color-mix.

Re-works color parsing to completely split resolution from consuming,
allowing for a single implementation of the parsing code, all returning
CSSUnresolvedColor values. This allows us to delay resolving relative
colors until needed, storing the components first as a CSSUnresolvedRelativeColor
and then a StyleRelativeColor (mirroring CSSUnresolvedColorMix and
StyleColorMix).

* LayoutTests/TestExpectations:
    - Removes skips for now passing currentColor tests.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color.html:
    - Added tests for currentColor at computed value time.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color.html:
    - Updated tests to match spec (and new implementation) which preserves
      declared values for declared value serialization of relative colors.

* LayoutTests/imported/w3c/web-platform-tests/css/support/color-testcommon.js:
    - Add helper which allows fuzzy testing using a specific property, rather
      than always hardcoding it to use "color".

* Source/WTF/wtf/VectorTraits.h:
    - Add UniqueRef to the set of smart pointers which get the simple class
      traits, needed to use UniqueRef with FixedVector.

* Source/WTF/wtf/text/TextStream.h:
(WTF::streamSizedContainer):
(WTF::operator<<):
    - Extract sized container serialization into a helper and use to replace
      duplicated implementations. Also add overload for FixedVector.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
    - Expose number serialization helper to allow identical serialization when
      not using CSSPrimitiveValue for numeric values.

* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::buildCSSTextRecursive):
    - Ensure that symbols get serialized when they are part of a calc().

* Source/WebCore/css/calc/CSSCalcSymbolNode.cpp:
* Source/WebCore/css/calc/CSSCalcSymbolNode.h:
(WebCore::CSSCalcSymbolNode::customCSSText const):
    - Expose serialization of a symbol node utilizing `nameLiteralForSerialization`.

* Source/WebCore/css/color/CSSAbsoluteColorResolver.h: Added.
    - Factor out code to resolve a parsed absolute color into a concrete Color.

* Source/WebCore/css/color/CSSColorConversion+ToColor.h: Added.
    - Factor out code to convert a "typed color" into a concrete Color, applying
      the rules specified by the descriptor provided.

* Source/WebCore/css/color/CSSColorConversion+ToTypedColor.h: Added.
    - Factor out code to convert a "parsed color" into a "typed color", normalizing
      values according to the color descriptor provided.

* Source/WebCore/css/color/CSSColorDescriptors.h:
    - Moves descriptors out of CSSPropertyParserHelpers, they are now used outside
      of that area.
    - Add additional helpers for working with the descriptors that compute the
      component types at various stages, such as when calc and raw symbols have
      not yet been evaluated.
    - Add additional per-descriptor data to support serialization of relative
      color functions that have not been fully resolved.

* Source/WebCore/css/color/CSSColorMixResolver.cpp: Renamed from Source/WebCore/css/color/CSSResolvedColorMix.cpp.
* Source/WebCore/css/color/CSSColorMixResolver.h: Renamed from Source/WebCore/css/color/CSSResolvedColorMix.h.
    - Renamed to follow the same pattern we are using with relative color resolution.

* Source/WebCore/css/color/CSSColorMixSerialization.cpp: Copied from Source/WebCore/css/color/CSSColorMixSerialization.h.
* Source/WebCore/css/color/CSSColorMixSerialization.h:
    - Moved a few non-template functions into an implementation file to avoid
      unnecessary inlining for non-perf critical code.

* Source/WebCore/css/color/CSSRelativeColorResolver.h: Added.
    - Factor out code to resolve a parsed relative color with its origin already
      resolved into a concrete Color. Resolution is performed in stages:

        1. If any components are raw symbols, replace them with a their value
           in the symbol table.
        2. If any components are `calc()` values, evaluate the `calc()` passing
           the symbol table for any symbolic values they retain.
        3. At this point, all the components are raw values of some type (number,
           percent, or angle), normalize those raw components into a typed color
           (e.g. SRGBA<float>, OKLab<float>, etc.).
        4. Wrap the typed color in the type-erased wrapper, Color.

* Source/WebCore/css/color/CSSRelativeColorSerialization.cpp: Added.
* Source/WebCore/css/color/CSSRelativeColorSerialization.h: Added.
    - Factor out code to serialize relative colors.

* Source/WebCore/css/color/CSSUnresolvedAbsoluteColor.cpp: Added.
* Source/WebCore/css/color/CSSUnresolvedAbsoluteColor.h: Added.
    - Type to represent fully resolved colors that are part of CSSUnresolvedColor tree.
      Currently used for absolute color values and the soon to be removed color-contrast().

* Source/WebCore/css/color/CSSUnresolvedColor.cpp:
* Source/WebCore/css/color/CSSUnresolvedColor.h:
    - Add all remaining color types to the value variant to represent declared color
      values. To retain the enough information for correct resolution of relative
      colors, the variant holds a fully specified CSSUnresolvedRelativeColor for each
      relative color supported. Hex and Keyword types were also added to allow removing
      use of CSSPrimitiveValue in the CSSUnresolvedColor tree. Now, child colors are
      represented as UniqueRef<CSSUnresolvedColor>.
    - The new createColor() function allows eager resolution of unresolved colors as
      needed by the `parseColorRaw()` functions that Canvas uses.

* Source/WebCore/css/color/CSSUnresolvedColorHex.cpp: Added.
* Source/WebCore/css/color/CSSUnresolvedColorHex.h: Added
    - Type to represent hex colors in the CSSUnresolvedColor tree.

* Source/WebCore/css/color/CSSUnresolvedColorKeyword.cpp: Added.
* Source/WebCore/css/color/CSSUnresolvedColorKeyword.h: Added.
    - Type to represent keyword colors in the CSSUnresolvedColor tree.

* Source/WebCore/css/color/CSSUnresolvedColorMix.cpp:
* Source/WebCore/css/color/CSSUnresolvedColorMix.h:
    - Re-worked to remove use of CSSPrimitiveValue representations of mix components,
      instead using `UniqueRef<CSSUnresolvedColor>` for the origin colors and
      `std::variant<PercentRaw, UnevaluatedCalc<PercentRaw>>` for the percentages.
      This preserves the type information and makes these usable in workers without
      extensive CSSValuePool piping.

* Source/WebCore/css/color/CSSUnresolvedColorResolutionContext.cpp: Added.
* Source/WebCore/css/color/CSSUnresolvedColorResolutionContext.h: Added.
    - Context used by the `parseColorRaw()` functions to resolve colors without
      going through the style system. Currently the context contains explicit
      values, but if pre-computing these turns out of be expensive, this can
      be converted into a delegate to allow lazy evaluation.

* Source/WebCore/css/color/CSSUnresolvedLightDark.cpp:
* Source/WebCore/css/color/CSSUnresolvedLightDark.h:
    - Like CSSUnresolvedColorMix, re-worked to remove use of CSSPrimitiveValue.

* Source/WebCore/css/color/CSSUnresolvedRelativeColor.cpp: Added.
* Source/WebCore/css/color/CSSUnresolvedRelativeColor.h: Added.
    - Type to represent relative color in the CSSUnresolvedColor tree.
    - Uses a color descriptor to define component types and stores the
      origin as a UniqueRef<CSSUnresolvedColor>.
    - Converts to the new StyleRelativeColor at style resolution time
      allowing a currentColor origin value to be preserved as late as
      possible.

* Source/WebCore/css/color/StyleColor.cpp:
* Source/WebCore/css/color/StyleColor.h:
    - Like CSSUnresolvedColor, the relative color types are added to the
      the kind variant, fully specified by color descriptor. One slightly
      unfortunate duplication is that all the relative colors require their
      own constructor functions in order to keep instantiation of the variant
      to the implementation file.
    - Like StyleColorMix, the StyleRelativeColor types are stored in UniqueRefs
      to avoid bloating the size for the relatively uncommon case of currentColor
      as part of the origin.
    - Also like StyleColorMix, the StyleRelativeColor types are first checked
      to see if they can be resolved, and if so, are stored as absolute colors
      instead.

* Source/WebCore/css/color/StyleColorMix.cpp:
* Source/WebCore/css/color/StyleColorMix.h:
    - Use PercentRaw for the Percentage to allow for clearer serialization.
    - Factor out check if the origin is absolute into a function to match
      relative color.
    - Sort functions to match other style color types.

* Source/WebCore/css/color/StyleCurrentColor.cpp:
* Source/WebCore/css/color/StyleCurrentColor.h:
    - Sort functions to match other style color types.

* Source/WebCore/css/color/StyleRelativeColor.h: Added.
    - Type to represent relative colors in the StyleColor tree.
    - Utilizes code shared with CSSUnresolvedRelativeColor for resolution
      and serialization.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.h:
    - Major rework of the parsing logic. Big takeaways:

        - Only does the consuming now, resolution has all been moved to
          other files.
        - Only one set of consumer functions, all working to return a
          CSSUnresolvedColor value. Now more separate normal vs raw
          distinction outside of exposed functions. The exposed functions
          consume the CSSUnresolvedColor, and resolve it to either
          a CSSPrimitiveValue or a raw Color.
        - ColorInterpolationMethod consuming was moved to its own
          file to continue to split things up.
        - Exposed functions have been merged where possible, with all
          now taking CSSColorParsingOptions, and all the raw functions
          also taking a CSSUnresolvedColorResolutionContext.
        - All the functions are now equally worker safe (or at least
          as worker safe as the one labeled worker safe used to be),
          so the worker safe exposed function has been removed.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorInterpolationMethod.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorInterpolationMethod.h: Added.
    Moved from CSSPropertyParserConsumer+Color.h/cpp.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+RawTypes.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+RawTypes.h:
    - Added serialization of raw types, used generically by the relative color
      and color mix serialization code.
    - Removed a number of FooOrBarRaw type definitions that were only used by
      the old color parsing code.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+UnevaluatedCalc.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UnevaluatedCalc.h:
    - Added serialization of UnevaluatedCalc, used generically by the relative color
      and color mix serialization code.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeBorderColor):
(WebCore::CSSPropertyParserHelpers::consumeFontPaletteValuesOverrideColors):
    - Switched to use the CSSColorParsingOptions when calling consumeColor allowing us
      to remove overload.

* Source/WebCore/css/process-css-properties.py:
    - Switched to use the CSSColorParsingOptions when calling consumeColor allowing us
      to remove overload.

* Source/WebCore/html/canvas/CanvasStyle.cpp:
    - Switched to use the plain parseColorRaw with "System" colors not in the allow set,
      rather than worker safe variant, since they are both the same now.
    - Removed hsl clamp opt setting, as that has been a no-op for a bit now.

* Source/WebCore/platform/graphics/ColorSerialization.cpp:
* Source/WebCore/platform/graphics/ColorSerialization.h:
    - Expose serialization of the color space names for use by relative color serialization.

* Source/WebCore/style/ColorFromPrimitiveValue.cpp:
* Source/WebCore/style/ColorFromPrimitiveValue.h:
    - Factor out conversion from CSS value keyword to StyleColor into its own function
      for use by CSSUnresolvedColorKeyword.

Canonical link: https://commits.webkit.org/278696@main
